libusb/os/darwin_usb.h

```diff
@@ -29,125 +29,132 @@
+#define MIN_INTERFACE_VERSION 650
+#elif (MAC_OS_X_VERSION_MIN_REQUIRED >= 100803)
+#define MIN_INTERFACE_VERSION 550
+#elif (MAC_OS_X_VERSION_MIN_REQUIRED >= 100407)
+#define MIN_INTERFACE_VERSION 245
+#else
+#define MIN_INTERFACE_VERSION 220
+#endif
+
+#define IOINTERFACE0(darwin_interface, version) ((darwin_interface)->interface.interface ## version)
+#define IOINTERFACE(darwin_interface, version) IOINTERFACE0(darwin_interface, version)
+#define DEFAULT_IOINTERFACE(darwin_interface) IOINTERFACE(darwin_interface, MIN_INTERFACE_VERSION)
 
 /* IOUSBDeviceInterface */
 
-/* New in OS 10.9.0. */
+union darwin_device {
 #if defined (kIOUSBDeviceInterfaceID650)
-
-#define usb_device_t    IOUSBDeviceInterface650
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID650
-#define DeviceVersion 650
-
-/* New in OS 10.7.3 but can't test deployment target to that granularity, so round up. */
-#elif defined (kIOUSBDeviceInterfaceID500)
-
-#define usb_device_t    IOUSBDeviceInterface500
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID500
-#define DeviceVersion 500
-
-/* New in OS 10.5.4 but can't test deployment target to that granularity, so round up. */
-#elif defined (kIOUSBDeviceInterfaceID320)
-
-#define usb_device_t    IOUSBDeviceInterface320
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID320
-#define DeviceVersion 320
-
-/* New in OS 10.5.0. */
-#elif defined (kIOUSBDeviceInterfaceID300)
-
-#define usb_device_t    IOUSBDeviceInterface300
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID300
-#define DeviceVersion 300
-
-/* New in OS 10.4.5 (or 10.4.6?) but can't test deployment target to that granularity, so round up. */
-#elif defined (kIOUSBDeviceInterfaceID245)
-
-#define usb_device_t    IOUSBDeviceInterface245
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID245
-#define DeviceVersion 245
-
-/* New in OS 10.2.3 but can't test deployment target to that granularity, so round up. */
-#elif defined (kIOUSBDeviceInterfaceID197)
-
-#define usb_device_t    IOUSBDeviceInterface197
-#define DeviceInterfaceID kIOUSBDeviceInterfaceID197
-#define DeviceVersion 197
-
+  /* New in OS 10.9.0. */
+  IOUSBDeviceInterface650 **device650;
+#endif
+#if defined (kIOUSBDeviceInterfaceID500)
+  /* New in OS 10.7.3 but can't test deployment target to that granularity, so round up. */
+  IOUSBDeviceInterface500 **device500;
+#endif
+#if defined (kIOUSBDeviceInterfaceID320)
+  /* New in OS 10.5.4 but can't test deployment target to that granularity, so round up. */
+  IOUSBDeviceInterface320 **device320;
+#endif
+#if defined (kIOUSBDeviceInterfaceID300)
+  /* New in OS 10.5.0. */
+  IOUSBDeviceInterface300 **device300;
+#endif
+#if defined (kIOUSBDeviceInterfaceID245)
+  /* New in OS 10.4.7. */
+  IOUSBDeviceInterface245 **device245;
+#endif
+  /* New in OS 10.2.3 but can't test deployment target to that granularity, so round up. Oldest
+   * version supported. */
+  IOUSBDeviceInterface197 **device197;
+};
+typedef union darwin_device usb_device_t;
```

Comment:
because the value of the `usb_device_t` is always casted to the "currently required" type, and most of the time only a "default" type is used, I believe it would be much more convenient to typedef a `usb_device_t` to a MIN_DEVICE_VERSION, so in most places it can be used as is w/o any cast, and use a macro or an explicit cast only on those few exceptions there are in the code

same regarding the interface interface

Reply:
That is a good point. Will make that change.

Reply:
> That is a good point. Will make that change.

@hjelmn gentle reminder :)

Reply:
Done.

Reply:
Great! Since I am queuing up some merges now, please let me merge this afterwards and I'll take care of the nanoversioning.